solarwinds_apm/oboe/oboe_sampler.py

```diff
@@ -0,0 +1,702 @@
+
+    @property
+    def headers(self) -> RequestHeaders:
+        return self._headers
+
+    @headers.setter
+    def headers(self, value: RequestHeaders):
+        self._headers = value
+
+    @property
+    def trace_options(self) -> TraceOptionsWithResponse | None:
+        return self._trace_options
+
+    @trace_options.setter
+    def trace_options(self, value: TraceOptionsWithResponse | None):
+        self._trace_options = value
+
+    def __str__(self):
+        return (
+            f"SampleState{{decision={self.decision}, "
+            f"attributes={self.attributes}, "
+            f"settings={self.settings}, "
+            f"trace_state={self.trace_state}, "
+            f"headers={self.headers}, "
+            f"trace_options={self.trace_options}}}"
+        )
+
+
+def _span_type(parent_span: Span | None = None) -> SpanType:
+    """
+    Determine the type of span based on the parent span.
+    """
+    parent_span_context = (
+        parent_span.get_span_context() if parent_span else None
+    )
+    if parent_span_context is None or not parent_span_context.is_valid:
+        return SpanType.ROOT
+    if parent_span_context.is_remote:
+        return SpanType.ENTRY
+    return SpanType.LOCAL
+
+
+class OboeSampler(Sampler, ABC):
+    def __init__(self, meter_provider: MeterProvider, logger: logging.Logger):
+        self._logger = logger
+        self._counters = Counters(meter_provider=meter_provider)
+        self._buckets = {
+            BucketType.DEFAULT: _TokenBucket(),
+            BucketType.TRIGGER_RELAXED: _TokenBucket(),
+            BucketType.TRIGGER_STRICT: _TokenBucket(),
+        }
+        self._settings: Settings | None = None
+
+    @property
+    def logger(self):
+        return self._logger
+
+    @logger.setter
+    def logger(self, new_logger):
+        self._logger = new_logger
+
+    @property
+    def counters(self):
+        return self._counters
+
+    @counters.setter
+    def counters(self, new_counters):
+        self._counters = new_counters
+
+    @property
+    def buckets(self):
+        return self._buckets
+
+    @buckets.setter
+    def buckets(self, new_buckets):
+        self._buckets = new_buckets
+
+    @property
+    def settings(self):
+        return self._settings
+
+    @settings.setter
+    def settings(self, new_settings):
+        self._settings = new_settings
+
+    @override
+    def should_sample(
+        self,
+        parent_context: "Context" | None,
+        trace_id: int,
+        name: str,
+        kind: SpanKind | None = None,
+        attributes: Attributes = None,
+        links: Sequence["Link"] | None = None,
+        trace_state: "TraceState" | None = None,
+    ) -> "SamplingResult":
+        parent_span = get_current_span(parent_context)
+        span_type = _span_type(parent_span)
+        self.logger.debug(f"span type is {SpanType(span_type).name}")
+
+        if span_type == SpanType.LOCAL:
+            return self._handle_local_span(parent_span)
```

Comment:
We could save on re-implementing these checks if `OboeSampler` instead inherits from [OTel SDK's ParentBased sampler](https://github.com/open-telemetry/opentelemetry-python/blob/dd00ddcce09208abf18c0d178996e855ed06f944/opentelemetry-sdk/src/opentelemetry/sdk/trace/sampling.py#L339-L374), similar to old [ParentBasedSwSampler](https://github.com/solarwinds/apm-python/blob/4caf5852935f3b32bdccf0bf1206f5132bd16291/solarwinds_apm/sampler.py#L620-L645). Then later we can do a simpler replacement of [default sampler](https://github.com/solarwinds/apm-python/blob/4caf5852935f3b32bdccf0bf1206f5132bd16291/solarwinds_apm/configurator.py#L93).

Reply:
Yes, apm-js implemented that together.

Reply:
Created a new ticket for later, so resolving this.